tools/elf2efi.py

```diff
@@ -0,0 +1,571 @@
+    # These sections are either needed during conversion to PE or are otherwise not needed
+    # in the final PE image.
+    IGNORE_SECTIONS = [
+        ".ARM.exidx",
+        ".dynamic",
+        ".dynstr",
+        ".dynsym",
+        ".eh_frame_hdr",
+        ".eh_frame",
+        ".gnu.hash",
+        ".hash",
+        ".note.gnu.build-id",
+        ".rel.dyn",
+        ".rela.dyn",
+    ]
+
+    # Known sections we care about and want to be in the final PE.
+    COPY_SECTIONS = [
+        ".data",
+        ".osrel",
+        ".rodata",
+        ".sbat",
+        ".sdmagic",
+        ".text",
+    ]
+
+    # By only dealing with allocating sections we effectively filter out debug sections.
+    if not elf_s["sh_flags"] & SH_FLAGS.SHF_ALLOC:
+        return False
+
+    if elf_s.name in IGNORE_SECTIONS:
+        return False
+
+    # For paranoia we only handle sections we know of. Any new sections that come up should
+    # be added to IGNORE_SECTIONS/COPY_SECTIONS and/or the linker script.
+    if elf_s.name not in COPY_SECTIONS:
+        raise RuntimeError(f"Unknown section {elf_s.name}, refusing.")
+
+    if elf_s["sh_addr"] % SECTION_ALIGNMENT != 0:
+        raise RuntimeError(f"Section {elf_s.name} is not aligned.")
+    if len(elf_s.name) > 8:
+        raise RuntimeError(f"ELF section name {elf_s.name} too long.")
+
+    return True
+
+
+def convert_elf_section(elf_s: ELFSection) -> PeSection:
+    pe_s = PeSection()
+    pe_s.Name = elf_s.name.encode()
+    pe_s.VirtualSize = elf_s.data_size
+    pe_s.VirtualAddress = elf_s["sh_addr"]
+    pe_s.SizeOfRawData = align_to(elf_s.data_size, FILE_ALIGNMENT)
+    pe_s.data = bytearray(elf_s.data())
+
+    if elf_s["sh_flags"] & SH_FLAGS.SHF_EXECINSTR:
+        pe_s.Characteristics = 0x60000020  # CNT_CODE|MEM_READ|MEM_EXECUTE
+    elif elf_s["sh_flags"] & SH_FLAGS.SHF_WRITE:
+        pe_s.Characteristics = 0xC0000040  # CNT_INITIALIZED_DATA|MEM_READ|MEM_WRITE
+    else:
+        pe_s.Characteristics = 0x40000040  # CNT_INITIALIZED_DATA|MEM_READ
+
+    return pe_s
+
+
+def copy_sections(elf: ELFFile, opt: PeOptionalHeader) -> list[PeSection]:
+    sections = []
+
+    for elf_s in elf.iter_sections():
+        if not use_section(elf_s):
+            continue
+
+        pe_s = convert_elf_section(elf_s)
+        if pe_s.Name == b".text":
+            opt.BaseOfCode = pe_s.VirtualAddress
+            opt.SizeOfCode += pe_s.VirtualSize
+        else:
+            opt.SizeOfInitializedData += pe_s.VirtualSize
+
+        if pe_s.Name == b".data" and isinstance(opt, PeOptionalHeader32):
+            opt.BaseOfData = pe_s.VirtualAddress
+
+        sections.append(pe_s)
+
+    return sections
+
+
+def apply_elf_relative_relocation(
+    reloc: ElfRelocation, image_base: int, sections: list[PeSection], addend_size: int
+):
+    [target] = [
+        pe_s
+        for pe_s in sections
+        if pe_s.VirtualAddress
+        <= reloc["r_offset"]
+        < pe_s.VirtualAddress + len(pe_s.data)
```

Comment:
Please format it in the usual fashion:
```py
[target] = [pe_s for pe_s in sections
            if pe_s.VirtualAddress <= reloc["r_offset"] < pe_s.VirtualAddress + len(pe_s.data)]
```

Reply:
I'd really rather not. I was using black the whole time and it's such a nice thing to not have to care about formatting. I can add a `fmt: on/off` if you really wanna insist. (The other argparse case I can force by adding a trailing comma).

Reply:
As written now, it's completely unreadable. It's fine to use black, but in cases where it makes a total mess, it must be disabled or worked around.